core/chains/evm/logpoller/log_poller.go

```diff
@@ -1233,17 +1294,325 @@ func (lp *logPoller) fillRemainingBlocksFromRPC(
+	return hex.EncodeToString(eventsTopicsKey)
+}
+
+func compareBlockNumbers(n, m *big.Int) (cmp int) {
+	if n != nil && m != nil {
+		return int(m.Uint64() - n.Uint64())
+	}
+	if n == nil {
+		cmp--
+	}
+	if m == nil {
+		cmp++
+	}
+	return cmp
+}
+
+// Exposes ethGetLogsReqs to tests, casting the results to []GetLogBatchElem and sorting them to make
+// the output more predictable and convenient for assertions
+func (lp *logPoller) EthGetLogsReqs(fromBlock, toBlock *big.Int, blockHash *common.Hash) []GetLogsBatchElem {
+	rawReqs := lp.ethGetLogsReqs(fromBlock, toBlock, blockHash)
+	reqs := make([]GetLogsBatchElem, len(rawReqs))
+	for i := range rawReqs {
+		reqs[i] = GetLogsBatchElem(rawReqs[i])
+	}
+
+	slices.SortStableFunc(reqs, func(a, b GetLogsBatchElem) int {
+		nilA, nilB := a.BlockHash() == nil, b.BlockHash() == nil
+		if nilA && !nilB {
+			return -1
+		} else if !nilA && nilB {
+			return 1
+		}
+		if !nilB && !nilA {
+			if cmp := bytes.Compare(a.BlockHash()[:], b.BlockHash()[:]); cmp != 0 {
+				return cmp
+			}
+		}
+
+		if cmp := compareBlockNumbers(a.FromBlock(), b.FromBlock()); cmp != 0 {
+			return cmp
+		}
+
+		if cmp := compareBlockNumbers(a.ToBlock(), b.ToBlock()); cmp != 0 {
+			return cmp
+		}
+
+		addressesA, addressesB := a.Addresses(), b.Addresses()
+		if len(addressesA) != len(addressesB) {
+			return len(addressesA) - len(addressesB)
+		}
+		for i := range addressesA {
+			if cmp := bytes.Compare(addressesA[i][:], addressesB[i][:]); cmp != 0 {
+				return cmp
+			}
+		}
+
+		topicsA, topicsB := a.Topics(), b.Topics()
+		if len(topicsA) != len(topicsB) { // should both be 4, but may as well handle more general case
+			return len(topicsA) - len(topicsB)
+		}
+		for i := range topicsA {
+			if len(topicsA[i]) != len(topicsB[i]) {
+				return len(topicsA[i]) - len(topicsB[i])
+			}
+			for j := range topicsA[i] {
+				if cmp := bytes.Compare(topicsA[i][j][:], topicsB[i][j][:]); cmp != 0 {
+					return cmp
+				}
+			}
+		}
+		return 0 // a and b are identical
+	})
+	return reqs
+}
+
+// The topics passed into RegisterFilter are 2D slices backed by arrays that the
+// caller could change at any time. We must have our own deep copy that's
+// immutable and thread safe. Similarly, we don't want to pass our mutex-protected
+// copy down the stack while sending batch requests and waiting for responses
+func copyTopics(topics [][]common.Hash) (clone [][]common.Hash) {
+	clone = make([][]common.Hash, len(topics))
+	for i, topic := range topics {
+		clone[i] = make([]common.Hash, len(topic))
+		copy(clone[i], topics[i])
+	}
+	return clone
+}
+
+// ethGetLogsReqs generates a batched rpc reqs for all logs matching registered filters,
+// copying cached reqs and filling in block range/hash if none of the registered filters have changed
+func (lp *logPoller) ethGetLogsReqs(fromBlock, toBlock *big.Int, blockHash *common.Hash) []rpc.BatchElem {
+	lp.filtersMu.Lock()
+
+	if len(lp.removedFilters) != 0 || len(lp.newFilters) != 0 {
```

Comment:
Why don't we rebuild these on insert/delete?

Reply:
I think the two main advantages of doing it this way (at least before this PR) were:
1. Optimization: if lots of filters are registered or unregistered in rapid succession (eg through CLO adding or removing a batch of jobs, or a single job like LogTriggers registering/unregistering a large # of user-defined filters in a loop), the cache* only needs to be rebuilt once, not once for every filter
2. Only one thread (the runloop) ever reads or writes to the cache, avoiding the need for worrying about any concurrency issues there. Only reading and writing of `lp.filters` needs to be protected by a mutex.

Also, not sure if this one matters but:
3. Filter registration can return quickly in most cases, rather than taking a while.

However, because I've also included some improvements in the cache reconstruction process (it no longer rebuilds it from scratch when a filter is removed), the first advantage has been reduced I guess. Removing a single filter can still cause it to have to rebuild the req's for lots of other filters: it rebuilds any reqs including filters which overlap in an way with the one being removed. But most of them will presumably be untouched, and for adding new filters it doesn't need to mess with anything unrelated to the added filters.

For 2, originally I was hoping we could just unlock `lp.filtersDirty` soon after entering into `ethGetLogsReqs()`, and have it unlocked for most of it while it messes with the cache. That turned out to be too difficult because there are so many interactions between the two, so I just left it locked throughout the whole cache reconstruction process. But I think this is still a big advantage in that it is unlocked during the actual rpc requests.  If there were other threads which could change the cache, then they could all end up getting blocked on rpc requests or vice versa.

(*) To be clear, by cache I mean what was called `lp.cachedAddresses` and `lp.cachedEventSigs` prior to this PR, and now becomes `lp.cachedReqsByAddress` and
`lp.cachedReqsByEventTopicsKey`

Reply:
Basically, the idea is to have rpc requests sent in the background, and not have that interfere with any of the calls down to log poller... whether it be registering & unregistering filters or querying for things. Changing this part of the architecture of LogPoller so that the two are interconnected and have to wait for each other seems like a big change, so I wouldn't make it unless there were a clear advantage of doing so.

Reply:
> Only one thread (the runloop) ever reads or writes to the cache, avoiding the need for worrying about any concurrency issues there. Only reading and writing of lp.filters needs to be protected by a mutex.

Are you sure about that? It seems that runLoop is the one entry point as you mentioned. However, you can spawn concurrent routine from `AsyncReplay -> Replay -> backfill -> batchFetchLogs`
On the other hand everything is synchronized behind `lp.filtersMu` so there should not be any concurrent access I think

Reply:
Ah, good point.  I think the two reasons I gave are accurate in that... these were the reasons why the code was that way when I started working on the PR.  But I forgot about the AsyncReplay refactor, which happened after I'd finished this PR and moved on to other projects, while waiting it was waiting for review. ;-)